### PR TITLE
Invision Community OAuth Support

### PR DESCRIPTION
**Pull request by AffectedArc07**
:cl: Core
Adds Invision Community as an OAuth Provider.
/:cl:

:cl: HTTP API
InvisionCommunity has been added to the OAuthProviders enum.
/:cl:

:cl: Configuration
**The new configuration version is 4.2.0**.
InvisonCommunity can now be configured as an OAuth provider. Takes ClientId, ClientSecret, RedirectUrl and ServerUrl as strings.
/:cl:

TODO:
- [x] Test code
- [x] Handle appropriate API version bumps
- [ ] Update webpanel after this is merged (https://github.com/tgstation/tgstation-server-webpanel/pull/99)

### CODE REVIEW

**Comment by AffectedArc07** [comment]
chrome_WG2HKGvv5e

File: src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.Extensions.Logging;
+
+using Tgstation.Server.Api.Models;
+using Tgstation.Server.Host.Configuration;
+using Tgstation.Server.Host.System;
+
+namespace Tgstation.Server.Host.Security.OAuth
+{
+	/// <summary>
+	/// OAuth validator for Discord.
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:13]: Discord

File: src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	/// </summary>
+	sealed class InvisionCommunityOAuthValidator : GenericOAuthValidator
+		{
+		/// <inheritdoc />
+		public override OAuthProvider Provider => OAuthProvider.InvisionCommunity;
+
+		/// <inheritdoc />
+		protected override Uri TokenUrl => new Uri($"{BaseProtocolPath}/oauth/token/");
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:21]:
```suggestion
		protected override Uri TokenUrl => new Uri($"{BaseProtocolPath}/oauth/token");
```

File: src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+		/// <inheritdoc />
+		protected override Uri UserInformationUrl => new Uri($"{BaseProtocolPath}/api/core/me");
+
+		/// <summary>
+		/// Base path to the server's OAuth endpoint.
+		/// </summary>
+		string BaseProtocolPath => $"{OAuthConfiguration.ServerUrl}";
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:29]: I can tell you copypasta'd the Keycloak provider, but this property is unecessary if it's just referencing the variable

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — CHANGES_REQUESTED** [review]
HTTP API and Configuration versions need a minor version bump

File: src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs (hunk 2 above)
Comment (AffectedArc07) [reply src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:21]: Nada
image

Comment (Cyberboss) [reply src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:21]: `System.Uri` handles the trailing slash so it's just a style thing

Comment (AffectedArc07) [reply src/Tgstation.Server.Host/Security/OAuth/InvisionCommunityOAuthValidator.cs:21]: Not a styling thing. It straight up breaks if you dont have it.

**Comment by Cyberboss** [comment]
Also needs HTTP API and Configuration 🆑 's

**Comment by AffectedArc07** [comment]
> Also needs HTTP API and Configuration 🆑 's



> HTTP API and Configuration versions need a minor version bump

Think I did both of these properly.

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**